### ISSUE TITLE
dyld: lazy symbol binding failed: Symbol not found: _SCDynamicStoreCopyProxies

### ISSUE THREAD

**Issue by ryandesign** [curl/curl#11502]
### I did this

Install curl 8.2.0 on OS X 10.8 (or on any Mac OS X or OS X version where curl was compiled with the macOS 10.11 SDK or earlier)

### I expected the following

curl compiles but fails to run, for example:

```
$ curl --version
dyld: lazy symbol binding failed: Symbol not found: _SCDynamicStoreCopyProxies
  Referenced from: /opt/local/lib/libcurl.4.dylib
  Expected in: flat namespace

dyld: Symbol not found: _SCDynamicStoreCopyProxies
  Referenced from: /opt/local/lib/libcurl.4.dylib
  Expected in: flat namespace

Trace/BPT trap: 5
```

This was reported to MacPorts here: https://trac.macports.org/ticket/67803

The problem seems to be that curl always wants to use `SCDynamicStoreCopyProxies` on macOS, and this is part of the SystemConfiguration framework, and has been ever since Mac OS X 10.1. However, curl only links with that framework when building against the macOS 10.12 SDK or later.

Normally a missing symbol would be a link-time error but by using the flags `-Wl,-undefined -Wl,dynamic_lookup` you're telling the linker to ignore the problem and that the symbol will exist at runtime. It doesn't, though, because you haven't linked in the framework that has the symbol.

This configure test is faulty:

```
checking whether to link macOS CoreFoundation and SystemConfiguration framework... no
```

It is checking whether `TARGET_OS_OSX` is defined. That define was newly added in the macOS 10.12 SDK.

It looks like an attempt to fix this issue was made in #11417 but it didn't fix it in m4/curl-sysconfig.m4.

`TARGET_OS_OSX` was introduced to help developers distinguish between macOS and other Darwin-derived operating systems (iOS, tvOS, watchOS). Is that distinction necessary here? If not, you can use `TARGET_OS_MAC` which encompasses all of them. Another way to identify any Darwin-based Apple OS is to check if both `__APPLE__` and `__MACH__` are defined.

### curl/libcurl version

8.2.0

### operating system

Darwin redacted 12.6.0 Darwin Kernel Version 12.6.0: Wed Mar 18 16:23:48 PDT 2015; root:xnu-2050.48.19~1/RELEASE_X86_64 x86_64

**Comment by ryandesign** [comment]
#11417 was merged after the 8.2.0 release so I haven't tested it but I'm still dubious of the use of `TARGET_OS_OSX` at all.

**Comment by ryandesign** [comment]
If the macOS distinction is important—if you need this code to happen on macOS but not iOS/tvOS/watchOS—a condition that was used in harfbuzz for that purpose was:

```c
#if TARGET_OS_MAC && !(defined(TARGET_OS_IPHONE) && TARGET_OS_IPHONE)
```

The now-not-particularly-accurately-named `TARGET_OS_IPHONE` means "any Darwin-derived OS that is not macOS". (It was introduced back when iOS, which was then called iPhoneOS, was the only non-macOS Darwin OS.)

**Comment by bagder** [comment]
> If the macOS distinction is important—if you need this code to happen on macOS but not iOS/tvOS/watchOS

We want it to work as widely as possible. But I don't know what that requires.

**Comment by ryandesign** [comment]
The question is what was the intent of #7121: to use it on any Darwin OS or only on macOS? It says:

> if `USE_RESOLVE_ON_IPS` is set (which is currently only set on macOS)

but looking at the code it is actually defined on all Apple platforms:

https://github.com/curl/curl/blob/57f56cc37459ade1dd86004a14c5493acc33ec60/lib/curl_setup.h#L258-L260

It also says:

> I need to check if it exists on targets like iOS

I `grep`'d through the SDKs and this function is mentioned in non-macOS SDKs:

```
% find /Volumes/BigSur/Applications/Xcode.app/Contents/Developer/Platforms -name 'SystemConfiguration.tbd' -or -name 'SCDynamicStoreCopySpecific.h' 2>/dev/null | xargs grep --color SCDynamicStoreCopyProxies            
/Volumes/BigSur/Applications/Xcode.app/Contents/Developer/Platforms/AppleTVOS.platform/Developer/SDKs/AppleTVOS.sdk/System/Library/Frameworks/SystemConfiguration.framework/SystemConfiguration.tbd:                       _SCDynamicStoreCopyProxies, _SCDynamicStoreCopyProxiesWithOptions, 
/Volumes/BigSur/Applications/Xcode.app/Contents/Developer/Platforms/AppleTVOS.platform/Developer/SDKs/AppleTVOS.sdk/System/Library/Frameworks/SystemConfiguration.framework/Headers/SCDynamicStoreCopySpecific.h:	@function SCDynamicStoreCopyProxies
/Volumes/BigSur/Applications/Xcode.app/Contents/Developer/Platforms/AppleTVOS.platform/Developer/SDKs/AppleTVOS.sdk/System/Library/Frameworks/SystemConfiguration.framework/Headers/SCDynamicStoreCopySpecific.h:SCDynamicStoreCopyProxies		(
/Volumes/BigSur/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk/System/Library/Frameworks/SystemConfiguration.framework/SystemConfiguration.tbd:                       _SCDynamicStoreCopyProxies, _SCDynamicStoreCopyProxiesWithOptions, 
/Volumes/BigSur/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk/System/Library/Frameworks/SystemConfiguration.framework/Headers/SCDynamicStoreCopySpecific.h:	@function SCDynamicStoreCopyProxies
/Volumes/BigSur/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk/System/Library/Frameworks/SystemConfiguration.framework/Headers/SCDynamicStoreCopySpecific.h:SCDynamicStoreCopyProxies		(
/Volumes/BigSur/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/SystemConfiguration.framework/SystemConfiguration.tbd:                       _SCDynamicStoreCopyNotifiedKeys, _SCDynamicStoreCopyProxies, 
/Volumes/BigSur/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/SystemConfiguration.framework/SystemConfiguration.tbd:                       _SCDynamicStoreCopyProxiesWithOptions, _SCDynamicStoreCopyValue, 
/Volumes/BigSur/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/SystemConfiguration.framework/Versions/A/SystemConfiguration.tbd:                       _SCDynamicStoreCopyNotifiedKeys, _SCDynamicStoreCopyProxies, 
/Volumes/BigSur/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/SystemConfiguration.framework/Versions/A/SystemConfiguration.tbd:                       _SCDynamicStoreCopyProxiesWithOptions, _SCDynamicStoreCopyValue, 
/Volumes/BigSur/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/SystemConfiguration.framework/Versions/A/Headers/SCDynamicStoreCopySpecific.h:	@function SCDynamicStoreCopyProxies
/Volumes/BigSur/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/SystemConfiguration.framework/Versions/A/Headers/SCDynamicStoreCopySpecific.h:SCDynamicStoreCopyProxies		(
/Volumes/BigSur/Applications/Xcode.app/Contents/Developer/Platforms/AppleTVSimulator.platform/Developer/SDKs/AppleTVSimulator.sdk/System/Library/Frameworks/SystemConfiguration.framework/SystemConfiguration.tbd:                       _SCDynamicStoreCopyProxies, _SCDynamicStoreCopyProxiesWithOptions, 
/Volumes/BigSur/Applications/Xcode.app/Contents/Developer/Platforms/AppleTVSimulator.platform/Developer/SDKs/AppleTVSimulator.sdk/System/Library/Frameworks/SystemConfiguration.framework/Headers/SCDynamicStoreCopySpecific.h:	@function SCDynamicStoreCopyProxies
/Volumes/BigSur/Applications/Xcode.app/Contents/Developer/Platforms/AppleTVSimulator.platform/Developer/SDKs/AppleTVSimulator.sdk/System/Library/Frameworks/SystemConfiguration.framework/Headers/SCDynamicStoreCopySpecific.h:SCDynamicStoreCopyProxies		(
/Volumes/BigSur/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/System/Library/Frameworks/SystemConfiguration.framework/SystemConfiguration.tbd:                       _SCDynamicStoreCopyProxies, _SCDynamicStoreCopyProxiesWithOptions, 
/Volumes/BigSur/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/System/Library/Frameworks/SystemConfiguration.framework/Headers/SCDynamicStoreCopySpecific.h:	@function SCDynamicStoreCopyProxies
/Volumes/BigSur/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/System/Library/Frameworks/SystemConfiguration.framework/Headers/SCDynamicStoreCopySpecific.h:SCDynamicStoreCopyProxies		(
```

But [this Stack Overflow post](https://stackoverflow.com/questions/42998035/scdynamicstorecreate-is-unavailable-not-available-on-ios) shows someone encountering these errors:

> ```
> 'SCDynamicStoreCreate' is unavailable: not available on iOS
> 'SCDynamicStoreCopyValue' is unavailable: not available on iOS
> ```

with someone responding:

> Those are only available for Mac OS

**Comment by bagder** [comment]
> what was the intent

I'm not sure that's terribly important. The *intent* of that PR was to fix the problem of the author, which I believe it did. That seems have been to make libcurl to behave better on macOS.

Does this logic also apply to other apple operating systems and versions or is it perhaps restricted to a subset? And if so, can the change be improved?

If so, I hope someone offers a PR that does it.  The intent of the PR from two years ago is no longer that interesting.

**Comment by ryandesign** [comment]
I tried a build for iOS after changing the conditions from `TARGET_OS_OSX` to `TARGET_OS_MAC` (which includes iOS and other Darwin OS variants) and the build failed:

```
macos.c:52:28: error: 'SCDynamicStoreCopyProxies' is unavailable: not available on iOS
    CFDictionaryRef dict = SCDynamicStoreCopyProxies(NULL);
                           ^
```

That confirms for me that this should really only be used on macOS, not other Darwin OS derivatives, and I'll prepare a PR to fix it so it'll work on older SDKs.

**Comment by ryandesign** [comment]
I haven't decided yet what conditional to use to identify macOS. Elsewhere in curl this conditional is used:

https://github.com/curl/curl/blob/3c0a91077c2af4a34f5ac177dd5ec8149ce479d7/lib/vtls/sectransp.c#L70

However Apple now says `TARGET_OS_EMBEDDED` is deprecated. I'm also keeping in mind the fact that newer clang compilers default to `-Werror,-Wundef-prefix=TARGET_OS_` and so you either have to be certain that the `TARGET_OS_` macros you use are defined on every OS or you have to check that they're `defined` first.

The conditional to identify macOS is currently repeated in four files. The breakage in 8.2.0 was caused by making a change to the conditional in only three out of the four files. Would it be better to define that conditional in fewer files so that it's less likely to go out of sync? I was thinking of defining it only here:

https://github.com/curl/curl/blob/3c0a91077c2af4a34f5ac177dd5ec8149ce479d7/m4/curl-sysconfig.m4#L33

however it would need to be duplicated in the CMake build system too. But then the configure script or CMake could define a macro in curl_config.h which the rest of the .c and .h files could use. If this is a good idea, what would be a good macro name? Are there other existing curl macros for identifying other operating systems?

**Comment by fds242** [comment]
Please consider including an easy escape hatch from getting this whole SCDynamicStoreCopyProxies call/code included if you're "fixing" the detection. I was depending on the happy accident of TARGET_OS_OSX not defined in older SDKs, and now due to the #11417 fix in curl 8.20 I had to start adding -DTARGET_OS_OSX=0 to CPPFLAGS avoid any of this code getting included, despite the build being for macOS.

As far as I'm concerned, this call is total poison, causing mysterious crashes making the culprit really hard to track down. For example, I had php-fpm immediately crashing on pdo_pgsql PostgreSQL database connection attempts, all due to PHP having an also loaded curl extension linked to libcurl with this SCDynamicStoreCopyProxies call included. And of course impossible to reproduce with a different PHP SAPI that does not use forking. For this reason, I do not want my macOS libcurl builds to actually include this call, ever.

**Comment by bagder** [comment]
> I do not want my macOS libcurl builds to actually include this call, ever.

Then your builds of libcurl will not have the same function on macOS as others' builds do, since this call is the only known way to enable that Ipv4 => IPv6 conversion thing.

**Comment by fds242** [comment]
> > I do not want my macOS libcurl builds to actually include this call, ever.
> Then your builds of libcurl will not have the same function on macOS as others' builds do, since this call is the only known way to enable that Ipv4 => IPv6 conversion thing.

That's an extremely esoteric case I cannot even imagine happening outside of some laboratory setting. An IPv6-only network interface, yet expecting to use plain global IPv4 addresses on it? Whereas, while shrinking, PHP still shows at 19.6% as the most-used non-C language binding in the user survey, and PHP-FPM is the default way of running PHP with web servers on any *nix platform. The majority of PHP / Mac users would surely prefer not having random inexplicable crashes to supporting such an impossibly rare use case.

Arguably, this is entirely the wrong call to include in a library in any case. Calling a function that, even in the happy case where it doesn't cause random crashes, is specifically meant to alter the behavior of other system calls (getaddrinfo) in the same process, should be left to the application, not any library. Ideally, this call should only be in curl, the command-line utility, and outside of libcurl. Then if any other libcurl-using code would wish to opt in to the same behavior, they can likewise enable it themselves for their entire process.

But I'm perfectly happy if I can just continue to keep my own libcurl builds crash-free, as it accidentally automatically was the case in curl < 8.20.0, and thankfully remains quite easy to do in 8.20.0. I'm only afraid of needlessly strengthening the macOS detection with no way out but manually patching the code going forwards, is all. Hence my original comment. Please think of it when making changes.
Others are quite welcome to have all their crashing builds at their leisure if that's their preference. It is quite natural that no two curl/libcurl builds are the same, and each is widely different depending on the list of third-party libraries linked in, TLS backend(s) chosen, features enabled, and so on.

**Comment by bagder** [comment]
> random inexplicable crashes

You are the first to report this problem, and you did not even submit an issue for it. You're piggy-backing a link issue. I'm not sure what to think about it and I'm certainly not convinced about this claim just with this.

**Comment by fds242** [comment]
> You are the first to report this problem, and you did not even submit an issue for it.

I have no issue or problem at present, since it is trivial to continue to build curl without this call included, without altering a single character of curl source code. I'm happy if things stay as is, unchanged. Except this "link issue" here is possibly going to result in changes to make such a build more difficult in the future, which I hoped to preempt by the comment. Hopefully this part of the code will remain easily disabled even post modifications for this issue, and not forced on all macOS builds, in which case I'll continue to have no issue of my own. Thank you!

**Comment by ryandesign** [comment]
I still think reducing the number of places the same conditional is repeated would be a good idea but for now I've submitted a PR that just fixes the conditionals.

I'll attempt to reproduce the PHP FPM problem and will file a separate issue if I find anything.

**Comment by bagder** [comment]
@fds242 if you can provide evidence of problems with this function call, preferably in a way to reproduce the issue, then we could work on fixing the root cause instead of you trying to disable parts of libcurl that is not written to be disabled easily. That would be great!

**Comment by bagder** [comment]
I'm not satisfied with that.

**Comment by ryandesign** [comment]
@fds242 I've attempted to use PHP FPM with the curl and pdo_pgsql modules and I haven't been able to make it crash, so I may not be testing the same way you were. Could you file a separate issue please with more information on your setup (macOS version, PHP version, which web server and version, a minimal reproduction recipe)?

**Comment by fds242** [comment]
@ryandesign Apologies for the off-topic reply here, but I simply don't know when I'd have the time to finish writing up & filing a proper separate issue myself. Thank you for wanting to look into it.

After too much trial-and-error, the key to triggering a 100% immediately reproducible crash is having a libpq (the postgresql client library) built with Kerberos support (`--with-gssapi`), loaded alongside curl containing the unfortunate call to SCDynamicStoreCopyProxies. You don't even have to try to invoke anything curl (since it now unconditionally executes the call on load), it simply has to be loaded to trigger a crash as soon as you attempt to connect to an actually existing PostgreSQL server. Only the version of curl is sensitive, all the relevant changes landed in 8.2.0. But everything I tested was the current non-beta release version. Web server is also irrelevant, Apache, Caddy, nginx all the same. I'm still on a x86_64 Mac. This I don't know if changes reproducibility. Compilers/SDKs/macOS version targets I tested a wide variety and made no difference.

Finally here are some draft build & test instructions that should make it easy to reproduce:
```
# no need for anything via Homebrew, MacPorts, or manual installs, only Xcode

# everything will just go here, easy to trash later
WORKDIR="$HOME/Desktop/crashtest"

mkdir "$WORKDIR" && cd "$WORKDIR"

# assuming you already downloaded them
tar -xjf ~/Downloads/curl-8.2.0.tar.bz2
tar -xjf ~/Downloads/postgresql-15.3.tar.bz2
tar -xjf ~/Downloads/php-8.2.8.tar.bz2

( cd curl-8.2.0/ && CPPFLAGS="-D TARGET_OS_OSX=1" ./configure --prefix="$WORKDIR" --without-ssl --disable-ldap && make -j4 install )

( cd postgresql-15.3/ && ./configure --prefix="$WORKDIR" --with-gssapi --without-ldap && make -j4 -C src/interfaces/libpq install && make -C src/include install )

( cd php-8.2.8/ && CURL_CFLAGS="-I$WORKDIR/include" CURL_LIBS=$("$WORKDIR"/bin/curl-config --libs) ./configure --prefix="$WORKDIR" --disable-all --enable-fpm --with-curl=shared,"$WORKDIR" --enable-pdo --with-pdo-pgsql=shared,"$WORKDIR" && make -j4 install )



## this keeps the default php-fpm listen setting of TCP localhost:9000, ensure nothing listens there already
mv -i etc/php-fpm.d/www.conf.default etc/php-fpm.d/www.conf
mv -i etc/php-fpm.conf.default etc/php-fpm.conf

## load both extensions
echo "extension = pdo_pgsql" > "$WORKDIR"/lib/php.ini
echo "extension = curl" >> "$WORKDIR"/lib/php.ini

## server has to actually exist (and support Kerberos?) to trigger the immediate crash, but valid working credentials don't matter
echo "<?php new PDO('pgsql:host=your-actual-working-postgresql-server');" > test.php

## the Caddy webserver is super easy to configure & test with
>Caddyfile cat <<-EOF
{
    http_port 9009
}
http:// {
    root * $WORKDIR
    php_fastcgi localhost:9000
}
EOF


# start PHP-FPM at last, this will keep this terminal occupied for easy debugging
"$WORKDIR/sbin/php-fpm" --nodaemonize

# get Caddy from https://github.com/caddyserver/caddy/releases
# make sure cwd is where we put the Caddyfile earlier
# this will also keep this terminal occupied
caddy run


# then you can finally test with for example:
curl -i http://localhost:9009/test.php

# you will get 502 Bad Gateway from Caddy when the php-fpm child crashes, and see the crash in php-fpm's output

# then see how the crash goes away if you stop loading the curl extension, or install a libcurl with the SCDynamicStoreCopyProxies call removed, etc.
# don't forget to restart php-fpm if changes were made, web server does not have to be restarted

```
